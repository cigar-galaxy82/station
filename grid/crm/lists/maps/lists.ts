/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable jest/no-export */

import { describe, expect } from '@jest/globals';
import { SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

export function getListsTest(providerName: string): void {
  describe(`crm/lists/get-lists/${providerName}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'crm/lists',
        provider: providerName,
      });
    });

    describe('GetLists', () => {
      it('returns all lists', async () => {
        const result = await superface.run({
          useCase: 'GetLists',
          input: {},
        });

        expect(() => result.unwrap()).not.toThrow();
        const data = result.unwrap();
        expect(data).toMatchSnapshot();
      });
    });
    // TODO: test pagination?
  });
}

export function addContactsToListTest(
  providerName: string,
  emails: string[],
  listId: string
): void {
  describe(`crm/lists/addContacts/${providerName}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'crm/lists',
        provider: providerName,
      });
    });

    describe('AddContactsToList', () => {
      describe('when all inputs are correct', () => {
        it('adds all contacts to a list', async () => {
          const result = await superface.run({
            useCase: 'AddContactsToList',
            input: {
              emails,
              listId,
            },
          });

          expect(() => result.unwrap()).not.toThrow();
          const data = result.unwrap();
          expect(data).toMatchSnapshot();
        });
        // TODO: ensure contacts were actually added
        // TODO: clean up after test
      });
    });
  });
}
