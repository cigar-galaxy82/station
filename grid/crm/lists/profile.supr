"""
Lists
Manage contact lists in Customer Relation Management systems.
"""

name = "crm/lists"
version = "1.0.0"

"""
Get All Lists
Get all available contact lists
"""
usecase GetLists safe {
  input {
    """
    Page
    Pagination cursor (obtained from `nextPage`).
    """
    page string
  }

  result {
    lists [
      {
        """
        ID
        Unique ID of the contact list.
        """
        listId! string!
        """
        Name
        Name or description of the contact list.
        """
        name string
        """
        Length
        Amount of contacts in the contact list.
        """
        length number
      }
    ]
    """
    Next page
    Next page cursor.
    """
    nextPage string
  }

  error Error
}

"""
Add Contacts to List
Adds existing contacts to an existing list.
"""
usecase AddContactsToList unsafe {
  input {
    """
    List ID
    Unique ID of the list to add a contact to.
    """
    listId! string!

    """
    Contact e-mails
    E-mail addresses of the contacts to add to the list.
    """
    emails! [string!]!
  }

  error Error
}

model Error {
  """
  Title
  A short, human-readable summary of the problem type.
  """
  title! string

  """
  Detail
  A human-readable explanation specific to this occurrence of the problem.
  """
  detail string
}