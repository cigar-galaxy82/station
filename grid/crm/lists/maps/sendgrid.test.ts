import { addContactsToListTest, getListsTest } from './lists';

getListsTest('sendgrid');
addContactsToListTest(
  'sendgrid',
  ['demo+added_one@superface.ai', 'demo+added_two@superface.ai'],
  '17a9df8c-9003-4c60-8e27-8492aebdacce',
);
