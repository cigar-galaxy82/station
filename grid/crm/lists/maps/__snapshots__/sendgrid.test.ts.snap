// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crm/lists/sendgrid GetLists GetLists returns all lists 1`] = `
Object {
  "lists": Array [
    Object {
      "length": 2,
      "listId": "17a9df8c-9003-4c60-8e27-8492aebdacce",
      "name": "test list",
    },
    Object {
      "length": 0,
      "listId": "761dc899-a2d6-40e2-82c6-312a3db2566c",
      "name": "test list two",
    },
  ],
  "nextPage": null,
}
`;
